server/s3_copy_object_action.cc

```diff
@@ -76,11 +77,21 @@ void S3CopyObjectAction::setup_steps() {
 void S3CopyObjectAction::get_source_bucket_and_object() {
   s3_log(S3_LOG_DEBUG, request_id, "%s Entry\n", __func__);
   std::string source = request->get_headers_copysource();
-  size_t separator_pos = source.find("/");
-  if (separator_pos != std::string::npos) {
-    source_bucket_name = source.substr(0, separator_pos);
-    source_object_name = source.substr(separator_pos + 1);
+  size_t separator_pos;
+  if (source[0] != '/') {
+    separator_pos = source.find("/");
+    if (separator_pos != std::string::npos) {
+      source_bucket_name = source.substr(0, separator_pos);
+      source_object_name = source.substr(separator_pos + 1);
+    }
+  } else {
+    separator_pos = source.find("/", 1);
+    if (separator_pos != std::string::npos) {
+      source_bucket_name = source.substr(1, separator_pos - 1);
+      source_object_name = source.substr(separator_pos + 1);
+    }
   }
+  source_object_name = evhttp_decode_uri(source_object_name.c_str());
```

Comment:
Is it really needed?

Reply:
Yes, we've seen from debug logs captured from Cloudberry testing that some special chars are not decoded, hence I've explicitly added this. Also, put-object API uses the same.

Reply:
What about memory leak?
http://www.wangafu.net/~nickm/libevent-1.4/doxygen/html/evhttp_8h.html#a681a8e967dec9f73fb8e27f465fdae4b

Reply:
`std::string source_object_name;`
Since this is on the stack, there is no memory leak. [Reference](https://stackoverflow.com/questions/4979681/how-to-properly-free-a-stdstring-from-memory)

Reply:
evhttp_decode_uri() returns a pointer to memory block which should be freed by the caller of the function

Reply:
Ok, will open new JIRA and take this investigation there.

Reply:
Opened bug : [EOS-18618](https://jts.seagate.com/browse/EOS-18618) for fixing memory leak at multiple places wherever we call evhttp_decode_uri.